### ISSUE TITLE
Question about MoLane dataset.

### ISSUE THREAD

**Issue by toshinyg** [juliangebele/CARLANE#5]
I would like to express my gratitude for your contribution to the unsupervised domain adaptation for lane detection.

I have a question regarding the MoLane dataset, specifically about why the false positives (FP) and false negatives (FN) have equal values. 

Thank you for your attention, and I look forward to your response.

**Comment by juliangebele** [comment]
Dear @toshinyg,

thank you for your appreciation concerning our paper. You are always welcome!

Regarding your question, I will first try to give you a high level intuition, and later a more mathematical explanation as to why we have the same number of FN & FP.

In a broad sense, False Positive and False Negative are defined as followed: 
- FP = the model predicted the lane(s) incorrectly
- FN = the model did non detect the lane(s)

Now imagine a case, where the model correctly classified the left lane, however it mispredicted the right lane. In this case, FP = FN = 1/2 (50%). The model did not predict the right lane (FN), but at the same time, it predicted a lane, that is not known as ground truth lane.

In a narrow sense, FP and FN as defined in the TuSimple Benchmark is defined as followed:

- FP = the number of wrong predicted lanes / the number of all predicted lanes
- FN = the number of missed ground-truth lanes in the predictions / the number of all ground-truth lanes

In the above case, we have 1 mispredicted lane and 2 predicted lanes in total, which is a FP of 1/2 = 50%, while also missing 1 ground truth lane and 2 ground truth lanes in total, which is 1/2 = 50% as well. 

This should hold true for other scenarios, as long as the number of predicted lanes is equal to the number of ground truth lanes.

I hope I could make it as clear as possible,

Best regards.

**Comment by toshinyg** [comment]
Dear @juliangebele,

I would like to express my gratitude for your kind explanation.

I have understood the example you mentioned.
However, I still have some questions remaining. When conducting my own experiments and visualizing the results, I observed cases where the model detected only one lane when there were two ground truth lanes in the MoLane dataset, or the model detected two lanes when there was only one ground truth lane.

In such cases, it appears that the number of ground truth lanes and the number of predicted lanes are different. So I could not understand why the overall false positives (FP) and false negatives (FN) are equal in MoLane dataset even if it includes these situations.

Thank you for your valuable time and assistance.

Sincerely,

**Comment by juliangebele** [comment]
This is in fact really interesting. In our benchmarks, we observed, that our trained models had different results on the TuSimple dataset concerning FP & FN, maybe due to the fact, that the number of ground truth lanes are often not equal to the number of predicted lanes. On MoLane however, the models seemed to detect the lanes consistently, which might lead to equal FP & FN. 

What model did you train and test? Is it one of which we mentioned in our paper benchmarks, or did you use a different one? I will try to take a closer look at the TuSimple benchmark this weekend and keep you informed, if I find some valuable information.

Best regards.

**Comment by toshinyg** [comment]
We modified the benchmark methods which you mentioned in the paper.
However, the visualization results on MoLane in the website (https://carlanebenchmark.github.io/) also showed the same situation (the number of ground truth lanes and the number of predicted lanes are different, but the overall false positives (FP) and false negatives (FN) are equal).
That is the reason why I feel difficult to fully understand.

Thank you for your valuable time and effort.
I look forward to your response.

Best wishes,

**Comment by juliangebele** [comment]
Yes, I see your point! I found something very unusual in the TuSimple benchmark metric, but for now I cannot tell, what part of it is wrong. I will dig deeper into the code and keep you updated on this!

All the best,

**Comment by toshinyg** [comment]
I'll think about it more, too.

Thank you for your time to this matter.
I look forward to your response.

Kind Regards,

**Comment by juliangebele** [comment]
Hi @toshinyg,

after some intense debugging, we finally found the error, which caused the FPs and FNs to be of the same value. 

If you take a look into our dataset labeling tool, non existing lanes are falsely saved as arrays, containing values such as [-2, -2, ..., -2]. In the TuSimple dataset, such lanes are not saved at all, in order to correctly perform the benchmark.

After finding the bug, we then wrote a python script to discard all of the pseudo-emtpy lanes in our .json files, containing only values such as -2, to mimic the TuSimple format. Then, we cleaned all .json files from MoLane, MuLane and the simulation files from TuLane, in order to recreate the entire benchmark for MoLane and MuLane, but didn't do this for TuLane, since the test files are already in the appropriate format. 

An important note, why we do not have to retrain the entire models for MoLane or MuLane: Before training, the .json labels are converted into .png grayscale images, which contain only the visible lanes. After this conversion, the .png labels are loaded into the PyTorch dataset class and a tensor is constructed, containing values from 0 - 100, where 0 - 99 indicates a valid lanepoint, while 100 indicates a missing lanepoint, which is similar to -2 in the .json files. All of this is done programmatically. As a result, our falsely saved pseudo-emtpy lanes like [-2, -2, ...] do not affect this process. However, since the .json files are used for the TuSimple benchmark, we only had to clean these labels. 

We observed, that MoLane now correctly contains different results for FP and FN. Hence, we will soon update the paper with the new benchmark results and upload the correct label files for each dataset as well as the code to discard the pseudo-empty lanes.

Thank you very much for your valuable comment and concern on this topic and your patience! We hope, we could clarify this issue well enough and shed light on that confusing bug! :) If there are any questions left on that topic, don't hesitate and always feel free to ask!

Best regards and a nice day,
